File: java/sys_nat/fan/sys/IntArray.java
package fan.sys;

import fanx.main.Sys;
import fanx.main.Type;

public abstract class IntArray {
	public static IntArray make(long size, long byteSize) {
		switch ((int)byteSize) {
		case 1:
			return new I1(size);
		case 2:
			return new I2(size);
		case 4:
			return new I4(size);
		case 8:
			return new I8(size);
		}
		throw UnsupportedErr.make("IntArray:"+byteSize);
	}

	public abstract long get(long pos);

	public abstract void set(long pos, long val);

	public abstract long size();
	
	public abstract Object raw();
	
	public IntArray fill(long val, long times) {
		int t = (int) times;
		for (int i = 0; i < t; ++i) {
			set(i, val);
		}
		return this;
	}

	public abstract IntArray realloc(long newSize);

	public abstract IntArray copyFrom(IntArray that, long thatOffset, long thisOffset, long length);
	
	public final Type typeof() { 
		if (type == null) type = Sys.findType("sys::IntArray");
		return type;
	}
	private static Type type;
	
	static class I4 extends IntArray {
		final int[] array;
		
		public I4(long size) {
			array = new int[(int)size];
		}
		
		public Object raw() {
			return array;
		}

		public long get(long pos) {
			return array[(int)pos];
		}

		public void set(long pos, long val) {
			array[(int)pos] = (int)val;
		}

		public long size() {
			return array.length;
		}

		public IntArray realloc(long newSize) {
			if (array.length == newSize) return this;
			I4 na = new I4(newSize);
			int len = array.length > na.array.length ? na.array.length : array.length;
			System.arraycopy(array, 0, na.array, 0, len);
			return na;
		}

		public I4 copyFrom(IntArray that, long thatOffset, long thisOffset, long length) {
			System.arraycopy(((I4)that).array, (int)thatOffset, array, (int)thisOffset, (int)length);
			return this;
		}
	}
	
	static class I8 extends IntArray {
		final long[] array;
		
		public I8(long size) {
			array = new long[(int)size];
		}
		
		public Object raw() {
			return array;
		}

		public long get(long pos) {
			return array[(int)pos];
		}

		public void set(long pos, long val) {
			array[(int)pos] = (long)val;
		}

		public long size() {
			return array.length;
		}

		public IntArray realloc(long newSize) {
			if (array.length == newSize) return this;
			I8 na = new I8(newSize);
			int len = array.length > na.array.length ? na.array.length : array.length;
			System.arraycopy(array, 0, na.array, 0, len);
			return na;
		}

		public I8 copyFrom(IntArray that, long thatOffset, long thisOffset, long length) {
			System.arraycopy(((I8)that).array, (int)thatOffset, array, (int)thisOffset, (int)length);
			return this;
		}
	}
	
	static class I2 extends IntArray {
		final short[] array;
		
		public I2(long size) {
			array = new short[(int)size];
		}
		
		public Object raw() {
			return array;
		}

		public long get(long pos) {
			return array[(int)pos];
		}

		public void set(long pos, long val) {
			array[(int)pos] = (short)val;
		}

		public long size() {
			return array.length;
		}

		public IntArray realloc(long newSize) {
			if (array.length == newSize) return this;
			I2 na = new I2(newSize);
			int len = array.length > na.array.length ? na.array.length : array.length;
			System.arraycopy(array, 0, na.array, 0, len);
			return na;
		}

		public I2 copyFrom(IntArray that, long thatOffset, long thisOffset, long length) {
			System.arraycopy(((I2)that).array, (int)thatOffset, array, (int)thisOffset, (int)length);
			return this;
		}
	}
	
	static class I1 extends IntArray {
		final byte[] array;
		
		public I1(long size) {
			array = new byte[(int)size];
		}
		
		public Object raw() {
			return array;
		}

		public long get(long pos) {
			return array[(int)pos];
		}

		public void set(long pos, long val) {
			array[(int)pos] = (byte)val;
		}

		public long size() {
			return array.length;
		}

		public IntArray realloc(long newSize) {
			if (array.length == newSize) return this;
			I1 na = new I1(newSize);
			int len = array.length > na.array.length ? na.array.length : array.length;
			System.arraycopy(array, 0, na.array, 0, len);
			return na;
		}

		public I1 copyFrom(IntArray that, long thatOffset, long thisOffset, long length) {
			System.arraycopy(((I1)that).array, (int)thatOffset, array, (int)thisOffset, (int)length);
			return this;
		}
	}
}

File: src/std/js/IntArray.js


fan.std.IntArray = fan.sys.Obj.$extend(fan.sys.Obj);
fan.std.IntArray.prototype.$ctor = function() {}
fan.std.IntArray.prototype.$typeof = function() { return fan.std.IntArray.$type; }

fan.std.IntArray.makeS1 = function(size) {
	self = new fan.std.IntArray()
	self.m_array = new Int8Array(size);
	return self
}

fan.std.IntArray.makeU1 = function(size) {
	self = new fan.std.IntArray()
	self.m_array = new Uint8Array(size);
	return self
}

fan.std.IntArray.makeS2 = function(size) {
	self = new fan.std.IntArray()
	self.m_array = new Int16Array(size);
	return self
}

fan.std.IntArray.makeU2 = function(size) {
	self = new fan.std.IntArray()
	self.m_array = new Uint16Array(size);
	return self
}

fan.std.IntArray.makeS4 = function(size) {
	self = new fan.std.IntArray()
	self.m_array = new Int32Array(size);
	return self
}

fan.std.IntArray.makeU4 = function(size) {
	self = new fan.std.IntArray()
	self.m_array = new Uint32Array(size);
	return self
}

fan.std.IntArray.makeS8 = function(size) {
	self = new fan.std.IntArray()
	self.m_array = new Int8Array(size);
	return self
}

fan.std.IntArray.prototype.get = function(pos) {
	return this.m_array[pos]
}

fan.std.IntArray.prototype.set = function(pos, val) {
	this.m_array[pos] = val
}

fan.std.IntArray.prototype.size = function() {
	return this.m_array.length;
}

fan.std.IntArray.prototype.realloc = function(newSize) {
	if (this.m_array.length == newSize) return this;

	if (newSize > this.m_array.length) {
		for (i = this.m_array.length; i<newSize; ++i) {
			this.m_array.push(null)
		}
		return this;
	}

	var na = fan.std.IntArray.make(newSize, this.m_of);
	var len = this.m_array.length > newSize ? newSize : this.m_array.length;
	for (var i = 0; i<len; ++i) {
		na.m_array[i] = this.m_array[i]
	}
	return na;
}

fan.std.IntArray.prototype.fill = function(val, times) {
	for (var i = 0; i < times; ++i) {
		this.m_array[i] = val;
	}
	return this;
}

fan.std.IntArray.prototype.copyFrom = function(that, thatOffset, thisOffset, length) {
	if (this === that) {
		this.m_array.copyWithin(thisOffset, thatOffset, thatOffset+length)
		return this;
	}

	for (var i = 0; i<length; ++i) {
		this.m_array[thisOffset + i] = that.m_array[i+thatOffset]
	}
	return this;
}

